### DISCUSSION TITLE
random 500 Internal Server Error - TypeError: fetch failed - in development mode

### DISCUSSION THREAD

**Discussion by MarkLyck** [vercel/next.js#55033] [Help]
### Summary

I've been building an application in Nextjs 13 with app router.

However every ~5-20 minutes or so when working on it in development mode, I'll refresh the page to see some updated content and I'll get a "500
Internal Server Error."

In the terminal it's printing this error when it happens:

```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async invokeRequest (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/server-ipc/invoke-request.js:17:12)
    at async invokeRender (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/router-server.js:254:29)
    at async handleRequest (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/router-server.js:447:24)
    at async requestHandler (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/router-server.js:464:13)
    at async Server.<anonymous> (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/start-server.js:117:13) {
  cause: Error: connect ECONNREFUSED ::1:49488
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1592:16) {
    errno: -61,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 49488
  }
}
```

When it gets in this state, I have to exit the current dev-mode and restart nextjs. It's incredibly frustrating doing this every 5-20 minutes.

It always happens after I refresh the browser (but most of the time when I refresh work fine without issues). It's seemingly random when it happens? Although I usually refresh the browser after I make a change in the code.

My current best guess is that that there's a small time-frame where if you make a code-change and nextjs is doing a hot-reload while you refresh the browser, nextJS gets into a bad state that it doesn't recover from? But I feel like I would be able to find other examples of this issue if that was the case?

This is what the network requests look like when it's happening:
<img width="1025" alt="Screenshot 2023-09-05 at 21 48 24" src="https://github.com/vercel/next.js/assets/6841110/fbebd5ed-e340-4db6-ae6f-1642066064a6">

It fails to fetch any resources from the nextjs server, going to another route doesn't fix it either.

Given the severe lack of information in this error, I don't even know where to start debugging this.

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
    Binaries:
      Node: 20.4.0
      npm: 9.7.2
      Yarn: 1.22.17
      pnpm: 8.7.0
    Relevant Packages:
      next: 13.4.19
      eslint-config-next: 13.4.19
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.2.2
    Next.js Config:
      output: N/A


Here's my next.js config:



/**
 * Run `build` or `dev` with `SKIP_ENV_VALIDATION` to skip env validation. This is especially useful
 * for Docker builds.
 */
await import('./env.mjs')

/** @type {import("next").NextConfig} */
const config = {
  reactStrictMode: false,
  experimental: {
    serverActions: true,
  },
  images: {
    remotePatterns: [
      {
        protocol: 'https',
        hostname: 'uploadthing.com',
        port: '',
      },
      {
        protocol: 'https',
        hostname: 'img.clerk.com',
        port: '',
      },
    ],
  },
}

export default config

```
```


### Example

_No response_

**Comment by Sakkhor909** [comment]
I am facing the same problem. But in my case even if I am not refreshing.
```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11576:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)   
    at async invokeRequest (D:\Personal\Work\Web Marketizer\Website\web-marketizer\node_modules\next\dist\server\lib\server-ipc\invoke-request.js:17:12)
    at async invokeRender (D:\Personal\Work\Web Marketizer\Website\web-marketizer\node_modules\next\dist\server\lib\router-server.js:254:29)
    at async handleRequest (D:\Personal\Work\Web Marketizer\Website\web-marketizer\node_modules\next\dist\server\lib\router-server.js:447:24)
    at async requestHandler (D:\Personal\Work\Web Marketizer\Website\web-marketizer\node_modules\next\dist\server\lib\router-server.js:464:13)
    at async Server.<anonymous> (D:\Personal\Work\Web Marketizer\Website\web-marketizer\node_modules\next\dist\server\lib\start-server.js:117:13) {
  cause: Error: connect ECONNREFUSED ::1:55698
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1495:16) {
    errno: -4078,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 55698
  }
}
```

**Reply by MarkLyck** [reply]
Please let me know if you solve it 🙏  been happening for weeks for me.

This guy also reported the same issue it looks like: https://github.com/vercel/next.js/discussions/54598

Another one with the same issue: https://github.com/vercel/next.js/discussions/55012

@leerob any idea what this is?

**Comment by Sakkhor909** [comment]
I am using React Icon. Maybe it causing the error. I don't know for sure. But I modularizeImports reac icon.
Read details here https://github.com/vercel/next.js/discussions/54359

After that I didn't face any problem. Also I install next@canary. Now everything is going smoothly.

**Reply by Stevemoretz** [reply]
Did you mean https://github.com/react-icons/react-icons/issues/786#issuecomment-1674964835 ?

**Reply by Sakkhor909** [reply]
Yes

**Reply by MarkLyck** [reply]
I'm not using react-icons, but it could be caused by multiple packages 🤷‍♂️

**Comment by Stevemoretz** [comment]
Thanks for that advice, I'm using the latest versions at the moment. I just
started using them and so far no issue, maybe they have fixed it in 13.4.20

```

***@***.***/all-files":
"https://github.com/react-icons/react-icons/releases/download/v4.11.0/react-icons-all-files-4.11.0.tgz",

"next": "^13.4.20-canary.35",

```

**Comment by ShahriarKh** [comment]
I'm facing this issue in prod! and really have no idea what causes this

**Comment by rcssdy** [comment]
I'm getting this issue locally. Have been searching for hours on how to fix it, no luck.

**Comment by thisisandreww** [comment]
Facing this issue in production intermittently, using og-image-generation.

**Comment by Mario-Beck** [comment]
Where does the port number in this error message come from?

**Comment by vladyslavKucheruk** [comment]
I'm facing the same error in the production environment. But when I refresh the page everything works as expected.
Is there a way to handle this "500 Internal Server Error." screen?

**Comment by Shapati** [comment]
> ### Summary
> I've been building an application in Nextjs 13 with app router.
> 
> However every ~5-20 minutes or so when working on it in development mode, I'll refresh the page to see some updated content and I'll get a "500 Internal Server Error."
> 
> In the terminal it's printing this error when it happens:
> 
> ```
> TypeError: fetch failed
>     at Object.fetch (node:internal/deps/undici/undici:11576:11)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async invokeRequest (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/server-ipc/invoke-request.js:17:12)
>     at async invokeRender (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/router-server.js:254:29)
>     at async handleRequest (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/router-server.js:447:24)
>     at async requestHandler (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/router-server.js:464:13)
>     at async Server.<anonymous> (/Users/marklyck/web/maker-pilot/node_modules/next/dist/server/lib/start-server.js:117:13) {
>   cause: Error: connect ECONNREFUSED ::1:49488
>       at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1592:16) {
>     errno: -61,
>     code: 'ECONNREFUSED',
>     syscall: 'connect',
>     address: '::1',
>     port: 49488
>   }
> }
> ```
> 
> When it gets in this state, I have to exit the current dev-mode and restart nextjs. It's incredibly frustrating doing this every 5-20 minutes.
> 
> It always happens after I refresh the browser (but most of the time when I refresh work fine without issues). It's seemingly random when it happens? Although I usually refresh the browser after I make a change in the code.
> 
> My current best guess is that that there's a small time-frame where if you make a code-change and nextjs is doing a hot-reload while you refresh the browser, nextJS gets into a bad state that it doesn't recover from? But I feel like I would be able to find other examples of this issue if that was the case?
> 
> This is what the network requests look like when it's happening: <img alt="Screenshot 2023-09-05 at 21 48 24" width="1025" src="https://private-user-images.githubusercontent.com/6841110/265866858-fbebd5ed-e340-4db6-ae6f-1642066064a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYxNzk4MjksIm5iZiI6MTcwNjE3OTUyOSwicGF0aCI6Ii82ODQxMTEwLzI2NTg2Njg1OC1mYmViZDVlZC1lMzQwLTRkYjYtYWU2Zi0xNjQyMDY2MDY0YTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjVUMTA0NTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjFjMjQ1OTIzODRhYWI2MjhhZDUzZmVjNDZlNDM3ZWJkZWRiMTJiMTNkMmE1ZmM0OTUzNTNhMDdjMDU2YWEwMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nXzO2Es3swlQw2PgrQbi19_KnCNRIm64Tnl1Opif9W4">
> 
> It fails to fetch any resources from the nextjs server, going to another route doesn't fix it either.
> 
> Given the severe lack of information in this error, I don't even know where to start debugging this.
> 
> ### Additional information
> ```js
> Operating System:
>       Platform: darwin
>       Arch: arm64
>       Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
>     Binaries:
>       Node: 20.4.0
>       npm: 9.7.2
>       Yarn: 1.22.17
>       pnpm: 8.7.0
>     Relevant Packages:
>       next: 13.4.19
>       eslint-config-next: 13.4.19
>       react: 18.2.0
>       react-dom: 18.2.0
>       typescript: 5.2.2
>     Next.js Config:
>       output: N/A
> 
> 
> Here's my next.js config:
> 
> 
> 
> /**
>  * Run `build` or `dev` with `SKIP_ENV_VALIDATION` to skip env validation. This is especially useful
>  * for Docker builds.
>  */
> await import('./env.mjs')
> 
> /** @type {import("next").NextConfig} */
> const config = {
>   reactStrictMode: false,
>   experimental: {
>     serverActions: true,
>   },
>   images: {
>     remotePatterns: [
>       {
>         protocol: 'https',
>         hostname: 'uploadthing.com',
>         port: '',
>       },
>       {
>         protocol: 'https',
>         hostname: 'img.clerk.com',
>         port: '',
>       },
>     ],
>   },
> }
> 
> export default config
> ```
> 
> ```
> 
> 
> ### Example
> 
> _No response_
> ```

any fix to this yet ?

**Comment by rodrigosebastiao** [comment]
I have 3 front+backs in production, it happens randomly, right now it killed one of my application, 3 days already.
Seems very specific between nextjs and strapi + hostinger, can't request to production API, fails in localhost and in vercel production.
But the APIs works fine everywhere else, insomnia/postman, direct browser by address or dev console.
https + SSH seems properly configured.

**Comment by giannif** [comment]
I'm seeing this in production. It was happening in about 1 - 20 refreshes, but then I upgraded to a canary (v14.3.0-canary.63) and the issue seemed to have gone away. Unfortunately 2 days later it showed up again.

`v14.3.0` does have some work done around externalizing `undici`, so hoping that fixes it. Here is a blog post that shares my frustrations, I have not tried their solution: https://medium.com/@kaloyan_17221/fix-vercel-next-js-fetch-failed-from-undici-polyfill-8c66346c9c2f

**Comment by princeajuzie7** [comment]
i  came across this issue on my next app and i solve this issue, by clearing my cache, deleting .next, node module and reinstall it again.

rm -rf node_modules
yarn cache clean
rm -rf .next
yarn install

hope this will help.
